processor/batchprocessor/config.go

```diff
@@ -1,68 +1,70 @@
+	// SendBatchMaxSize corresponds to the JSON schema field "send_batch_max_size".
 	SendBatchMaxSize uint32 `mapstructure:"send_batch_max_size"`
 
-	// MetadataKeys is a list of client.Metadata keys that will be
-	// used to form distinct batchers.  If this setting is empty,
-	// a single batcher instance will be used.  When this setting
-	// is not empty, one batcher will be used per distinct
-	// combination of values for the listed metadata keys.
-	//
-	// Empty value and unset metadata are treated as distinct cases.
-	//
-	// Entries are case-insensitive.  Duplicated entries will
-	// trigger a validation error.
-	MetadataKeys []string `mapstructure:"metadata_keys"`
+	// SendBatchSize corresponds to the JSON schema field "send_batch_size".
+	SendBatchSize uint32 `mapstructure:"send_batch_size"`
 
-	// MetadataCardinalityLimit indicates the maximum number of
-	// batcher instances that will be created through a distinct
-	// combination of MetadataKeys.
-	MetadataCardinalityLimit uint32 `mapstructure:"metadata_cardinality_limit"`
+	// Timeout corresponds to the JSON schema field "timeout".
+	Timeout time.Duration `mapstructure:"timeout"`
 }
 
-var _ component.Config = (*Config)(nil)
-
-// Validate checks if the processor configuration is valid
-func (cfg *Config) Validate() error {
-	if cfg.SendBatchMaxSize > 0 && cfg.SendBatchMaxSize < cfg.SendBatchSize {
-		return errors.New("send_batch_max_size must be greater or equal to send_batch_size")
+// UnmarshalJSON implements json.Unmarshaler.
+func (j *Config) UnmarshalJSON(value []byte) error {
+	var raw map[string]interface{}
+	if err := json.Unmarshal(value, &raw); err != nil {
+		return err
 	}
-	uniq := map[string]bool{}
-	for _, k := range cfg.MetadataKeys {
-		l := strings.ToLower(k)
-		if _, has := uniq[l]; has {
-			return fmt.Errorf("duplicate entry in metadata_keys: %q (case-insensitive)", l)
-		}
-		uniq[l] = true
+	type Plain Config
+	var plain Plain
+	if err := json.Unmarshal(value, &plain); err != nil {
+		return err
+	}
+	if v, ok := raw["metadata_cardinality_limit"]; !ok || v == nil {
+		plain.MetadataCardinalityLimit = 1000.0
+	}
+	if v, ok := raw["metadata_keys"]; !ok || v == nil {
+		plain.MetadataKeys = []string{}
+	}
+	if v, ok := raw["send_batch_max_size"]; !ok || v == nil {
+		plain.SendBatchMaxSize = 0.0
 	}
-	if cfg.Timeout < 0 {
-		return errors.New("timeout must be greater or equal to 0")
+	if v, ok := raw["send_batch_size"]; !ok || v == nil {
+		plain.SendBatchSize = 8192.0
 	}
+	if v, ok := raw["timeout"]; !ok || v == nil {
+		defaultDuration, err := time.ParseDuration("200ms")
+		if err != nil {
+			return fmt.Errorf("failed to parse the \"200ms\" default value for field timeout:%w }", err)
+		}
+		plain.Timeout = defaultDuration
+	}
+	*j = Config(plain)
 	return nil
 }
+
+// SetDefaults sets the fields of Config to their defaults.
+// Fields which do not have a default value are left untouched.
+func (c *Config) SetDefaults() {
```

Comment:
Hi @atoulme, would you be happy with an autogenerated `SetDefaults()` function such as this one? 

I think we will also need a `func (c *Config) Validate()` function. It could check for constraints such as whether a number is within a certain range. I'm not yet sure how (and if) it would check for required properties though.

cc @omissis